### PR TITLE
Remove some return-type diagnostic booleans from `FnCtxt`

### PR DESCRIPTION
**Pull request by compiler-errors**
These can be calculated on-demand

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/coercion.rs [hunk 1]
```diff
@@ -1782,7 +1782,8 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
             // may occur at the first return expression we see in the closure
             // (if it conflicts with the declared return type). Skip adding a
             // note in this case, since it would be incorrect.
-            && !fcx.return_type_pre_known
+            && let Some(fn_sig) = fcx.body_fn_sig()
+            && fn_sig.output().is_ty_var()
```
Comment (fee1-dead) [inline compiler/rustc_hir_typeck/src/coercion.rs:1786]: Shouldn't this be `is_ty_infer`?

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/coercion.rs:1786]: Doesn't really matter since users can't write numeric infer vars.

Comment (fee1-dead) [reply compiler/rustc_hir_typeck/src/coercion.rs:1786]: oh right that is true.

@bors r+

**Pull request merged**